### ISSUE TITLE
Feedback for switching from ids4

### ISSUE THREAD

**Issue by tai-yi** [openiddict/openiddict-core#1347]
I'm trying change my OIDC server core lib from ids4 to openiddcit （.net5).

Openiddcit is a very good lib to user extend their own logic, but there are still some limitations, so I have to replace a lot buildin handlers.

### OIDC configuration Issuer
`System.Uri` ToString will always include `/` when using root path (`https://a.com` -> `https://a.com/`).

I have to change it because I have some issued tokens in user clients. Right now I replaced many buildin handlers.

Can OpenIddict change Issuer type from `Uri` to `string` or priovide a configurable issuer formater?

### Refresh Keys

Keys renew is a very usual scenario. It is not very easy to implment keys automatically refresh in curren version.
Can Openiddcit provide an interface that allow user implment to provide SignInCredentials & JWKS & EncryptingCredentials (async function)

### Handlers Graph

Openiddcit has a very powerful pipeline model, but it also bring complexity, it is not fridenly for new guys.

I also suggest our log can write a sample graph to show how a context handled.

I write sample code localy. Output and code as below

Output
<details>

```
info: OpenIddict.Server.OpenIddictServerDispatcher.GraphWriter[0]
      |OpenIddict.Server.ProcessRequestContext
      |  >OpenIddict.Server.AspNetCore.InferEndpointType
      |  >OpenIddict.Server.AspNetCore.InferIssuerFromHost
      |  >OpenIddict.Server.ExtractAuthorizationRequest
      |  |OpenIddict.Server.ExtractAuthorizationRequestContext
      |  |  >OpenIddict.Server.AspNetCore.ExtractGetOrPostRequest`1
      |  >OpenIddict.Server.ValidateAuthorizationRequest
      |  |  |OpenIddict.Server.ValidateAuthorizationRequestContext
      |  |  |  >OpenIddict.Server.ValidateRequestParameter
      |  |  |  >OpenIddict.Server.ValidateRequestUriParameter
      |  |  |  >OpenIddict.Server.ValidateClientIdParameter
      |  |  |  >OpenIddict.Server.ValidateRedirectUriParameter
      |  |  |  >OpenIddict.Server.ValidateResponseTypeParameter
      |  |  |  >OpenIddict.Server.ValidateResponseModeParameter
      |  |  |  >OpenIddict.Server.ValidateScopeParameter
      |  |  |  >OpenIddict.Server.ValidateNonceParameter
      |  |  |  >OpenIddict.Server.ValidatePromptParameter
      |  |  |  >OpenIddict.Server.ValidateProofKeyForCodeExchangeParameters
      |  |  |  >OpenIddict.Server.ValidateClientId
      |  |  |  >OpenIddict.Server.ValidateClientType
      |  |  |  >OpenIddict.Server.ValidateClientRedirectUri
      |  |  |  >OpenIddict.Server.ValidateScopes
      |  |  |  >OpenIddict.Server.ValidateEndpointPermissions
      |  |  |  >OpenIddict.Server.ValidateGrantTypePermissions
      |  |  |  >OpenIddict.Server.ValidateResponseTypePermissions
      |  |  |  >OpenIddict.Server.ValidateScopePermissions
      |  |  |  >OpenIddict.Server.ValidateProofKeyForCodeExchangeRequirement
      |  >OpenIddict.Server.HandleAuthorizationRequest
      |  |  |  |OpenIddict.Server.HandleAuthorizationRequestContext
      |  |  |  |  >OpenIddict.Server.AspNetCore.EnablePassthroughMode`2
      |  |  |  |  -OpenIddict.Server.AspNetCore.EnablePassthroughMode`2 -- (skipped)
      |  -OpenIddict.Server.HandleAuthorizationRequest -- (skipped)
```
</details>

`OpenIddictServerDispatcher`
<details>

```C#
/*
 * Licensed under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
 * See https://github.com/openiddict/openiddict-core for more information concerning
 * the license and the contributors participating to this project.
 */

using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using static OpenIddict.Server.OpenIddictServerEvents;
using SR = OpenIddict.Abstractions.OpenIddictResources;

namespace OpenIddict.Server
{
    public class OpenIddictServerDispatcher : IOpenIddictServerDispatcher
    {
        private readonly ILogger<OpenIddictServerDispatcher> _logger;
        private readonly IOptionsMonitor<OpenIddictServerOptions> _options;
        private readonly IServiceProvider _provider;
        private readonly GraphWriter _writer;
        /// <summary>
        /// Creates a new instance of the <see cref="OpenIddictServerDispatcher"/> class.
        /// </summary>
        public OpenIddictServerDispatcher(
            ILoggerFactory loggerFactory,
            IOptionsMonitor<OpenIddictServerOptions> options,
            IServiceProvider provider)
        {
            _logger = loggerFactory.CreateLogger<OpenIddictServerDispatcher>();
            _options = options;
            _provider = provider;
            _writer = new GraphWriter(loggerFactory.CreateLogger<GraphWriter>());
        }

        private class GraphWriter : IDisposable
        {
            public GraphWriter(ILogger logger)
            {
                _logger = logger;
            }
            private readonly ILogger _logger;

            private StringBuilder _stringBuilder = new StringBuilder();
            private GraphWriter? _parent;
            public int Indent { get; set; }

            public void Dispose()
            {
                if (_parent == null)
                    return;
                _parent.Indent--;
                if (this._parent.Indent == 0)
                {
                    _logger.LogInformation(_stringBuilder.ToString());
                    _stringBuilder.Clear();
                }
                _parent = null;
            }

            public GraphWriter Create()
            {
                var graph = new GraphWriter(_logger)
                {
                    _parent = this,
                    _stringBuilder = _stringBuilder
                };
                Indent++;
                return graph;
            }

            public void WriteContext(Type type)
            {
                AppendIndent();
                _stringBuilder.Append('|');
                _stringBuilder.AppendLine(GetTypeName(type));
            }
            private void AppendIndent(int buf = 0)
            {
                for (int i = 0; i < Indent + buf; i++)
                {
                    _stringBuilder.Append("|  ");
                }
            }
            private static string? GetTypeName(Type type) => $"{type.Namespace}.{type.Name}";
            public void WriteHandler(Type type)
            {
                AppendIndent(1);
                _stringBuilder.Append('>');
                _stringBuilder.AppendLine(GetTypeName(type));
            }
            public void EndHandler(Type type, string? result)
            {
                AppendIndent(1);
                _stringBuilder.Append('-');
                _stringBuilder.AppendLine($"{GetTypeName(type)} -- ({result})");
            }
        }

        public async ValueTask DispatchAsync<TContext>(TContext context) where TContext : BaseContext
        {
            if (context is null)
            {
                throw new ArgumentNullException(nameof(context));
            }
            using var graph = _writer.Create();
            graph.WriteContext(typeof(TContext));
            await foreach (var handler in GetHandlersAsync())
            {
                graph.WriteHandler(handler.GetType());
                try
                {
                    await handler.HandleAsync(context);
                }

                catch (Exception exception) when (_logger.IsEnabled(LogLevel.Debug))
                {
                    graph.EndHandler(handler.GetType(), "error");
                    _logger.LogDebug(exception, SR.GetResourceString(SR.ID6132), handler.GetType().FullName, typeof(TContext).FullName);
                    throw;
                }

                if (_logger.IsEnabled(LogLevel.Debug))
                {
                    _logger.LogDebug(SR.GetResourceString(SR.ID6133), typeof(TContext).FullName, handler.GetType().FullName);
                }

                switch (context)
                {
                    case BaseRequestContext { IsRequestHandled: true }:
                        if (_logger.IsEnabled(LogLevel.Debug))
                        {
                            _logger.LogDebug(SR.GetResourceString(SR.ID6134), typeof(TContext).FullName, handler.GetType().FullName);
                        }
                        graph.EndHandler(handler.GetType(), "handled");
                        return;

                    case BaseRequestContext { IsRequestSkipped: true }:
                        if (_logger.IsEnabled(LogLevel.Debug))
                        {
                            _logger.LogDebug(SR.GetResourceString(SR.ID6135), typeof(TContext).FullName, handler.GetType().FullName);
                        }
                        graph.EndHandler(handler.GetType(), "skipped");
                        return;

                    case BaseValidatingContext { IsRejected: true }:
                        if (_logger.IsEnabled(LogLevel.Debug))
                        {
                            _logger.LogDebug(SR.GetResourceString(SR.ID6136), typeof(TContext).FullName, handler.GetType().FullName);
                        }
                        graph.EndHandler(handler.GetType(), "rejected");
                        return;

                    default: continue;
                }
            }

            async IAsyncEnumerable<IOpenIddictServerHandler<TContext>> GetHandlersAsync()
            {
                // Note: the descriptors collection is sorted during options initialization for performance reasons.
                var descriptors = _options.CurrentValue.Handlers;
                if (descriptors.Count == 0)
                {
                    yield break;
                }

                for (var index = 0; index < descriptors.Count; index++)
                {
                    var descriptor = descriptors[index];
                    if (descriptor.ContextType != typeof(TContext) || !await IsActiveAsync(descriptor))
                    {
                        continue;
                    }

                    var handler = descriptor.ServiceDescriptor.ImplementationInstance is not null ?
                        descriptor.ServiceDescriptor.ImplementationInstance as IOpenIddictServerHandler<TContext> :
                        _provider.GetService(descriptor.ServiceDescriptor.ServiceType) as IOpenIddictServerHandler<TContext>;

                    if (handler is null)
                    {
                        throw new InvalidOperationException(SR.FormatID0098(descriptor.ServiceDescriptor.ServiceType));
                    }

                    yield return handler;
                }
            }

            async ValueTask<bool> IsActiveAsync(OpenIddictServerHandlerDescriptor descriptor)
            {
                for (var index = 0; index < descriptor.FilterTypes.Length; index++)
                {
                    if (!(_provider.GetService(descriptor.FilterTypes[index]) is IOpenIddictServerHandlerFilter<TContext> filter))
                    {
                        throw new InvalidOperationException(SR.FormatID0099(descriptor.FilterTypes[index]));
                    }

                    if (!await filter.IsActiveAsync(context))
                    {
                        return false;
                    }
                }

                return true;
            }
        }
    }
}

```
</details>

**Comment by kevinchalet** [comment]
Thanks for your feedback.

> Can OpenIddict change Issuer type from `Uri` to `string` or priovide a configurable issuer formater?

Applications are encouraged to normalize URIs with an empty path to include a trailing slash (https://datatracker.ietf.org/doc/html/rfc3986#section-6.2.3). It's something all browsers systematically do. As such, I'm not super confortable adopting a breaking change just for that.

That said, 4.x will have new `GenerateToken`/`ValidateToken` events directly inspired from ASOS/OpenIddict 2.x's `Serialize*Token`/`Deserialize*Token` that you'll be able to use to override how tokens are generated, including how the `iss` claim is formatted.

> Openiddcit has a very powerful pipeline model, but it also bring complexity, it is not fridenly for new guys.
> I also suggest our log can write a sample graph to show how a context handled.

That's a good idea!

> Keys renew is a very usual scenario. It is not very easy to implment keys automatically refresh in curren version.
> Can Openiddcit provide an interface that allow user implment to provide SignInCredentials & JWKS & EncryptingCredentials (async function)

I agree it's a very frequent - but advanced and complicated - scenario. I'm seriously considering adding some key management love to OpenIddict, but it's very likely going to be a paid/sponsors-only thing, separate from the core, as it's a fairly big thing to implement.

**Comment by sam17896** [comment]
how frequent does the key renews, do we have exact time document somewhere ?

**Comment by kevinchalet** [comment]
> System.Uri ToString will always include / when using root path (https://a.com -> https://a.com/).
I have to change it because I have some issued tokens in user clients. Right now I replaced many buildin handlers.
Can OpenIddict change Issuer type from Uri to string or priovide a configurable issuer formater?

`Uri` is still used but OpenIddict 4.0 will specify both values when preparing the `TokenValidationParameters` sent to IdentityModel so having custom handlers will no longer be necessary.

Closing.
